File: NetFramework/DataFilter/VB/ChartAsDataFilter/Controls/FlexChartFilterView.vb
Imports C1.Chart
Imports C1.Win.Chart

Namespace ChartAsDataFilter
    Partial Public Class FlexChartFilterView
        Inherits UserControl
#Region "fields"
        Private _selectedIndices As List(Of Integer)
        Private _selecting As Boolean, _showSelectionMode As Boolean = True, _showRangeSlider As Boolean = False
        Private _ptStart As Point = Point.Empty, _ptEnd As Point = Point.Empty
        Private _items As IEnumerable(Of Object)
        Private _selectionMode As SelectionMode = SelectionMode.NormalSelect
        Private _chbClear As CheckBox
        Private _cbSelectionMode As ComboBox
        Private _chart As FlexChart
        Private _slider As RangeSliderView
        Protected _pnlControlOptions As FlowLayoutPanel
#End Region
#Region "Properties"
        Public ReadOnly Property Chart As FlexChart
            Get
                Return _chart
            End Get
        End Property

        Public ReadOnly Property Slider As RangeSliderView
            Get
                Return _slider
            End Get
        End Property

        Public Property SelectionMode As SelectionMode
            Get
                Return _selectionMode
            End Get
            Set(ByVal value As SelectionMode)

                If _selectionMode <> value Then
                    _selectionMode = value
                    _cbSelectionMode.SelectedIndex = If(_cbSelectionMode.SelectedIndex = -1, -1, CInt(value))
                End If
            End Set
        End Property

        Public Property ShowSelectionMode As Boolean
            Get
                Return _showSelectionMode
            End Get
            Set(ByVal value As Boolean)

                If value <> _showSelectionMode AndAlso _cbSelectionMode IsNot Nothing Then
                    _showSelectionMode = value
                    _cbSelectionMode.Visible = value
                End If
            End Set
        End Property

        Public Property ShowRangeSlider As Boolean
            Get
                Return _showRangeSlider
            End Get
            Set(ByVal value As Boolean)
                _showRangeSlider = value
                If Me.IsHandleCreated Then Slider.Visible = value
            End Set
        End Property

        Public ReadOnly Property IsGroupedData As Boolean
            Get
                Return _items.First().[GetType]() = GetType(GroupItem) OrElse _items.First().[GetType]().IsSubclassOf(GetType(GroupItem))
            End Get
        End Property

        Public ReadOnly Property IsFilterApplied As Boolean
            Get
                Return Not _chbClear.Checked
            End Get
        End Property
#End Region
#Region "Events"
        Public Event SelectionChanged As EventHandler
#End Region
#Region "c'tor"
        Public Sub New(ByVal items As IEnumerable(Of Object), ByVal bindingX As String, ByVal binding As String, ByVal Optional chartType As ChartType = ChartType.Column)
            InitializeComponent()
            _selectedIndices = New List(Of Integer)()
            Me._items = items
            _chart = New FlexChart() With {
                .ChartType = chartType,
                .DataSource = items,
                .BindingX = bindingX,
                .Binding = binding,
                .Dock = DockStyle.Fill,
                .Margin = New Padding(0, 30, 0, 0)
            }
            Dim ser = New Series()
            AddHandler ser.SymbolRendering, AddressOf OnSeriesSymbolRendering
            Me.Chart.Series.Add(ser)
            Me.Chart.SelectionStyle.Stroke = Brushes.DarkBlue
            Me.Chart.SelectionStyle.StrokeWidth = 2
            Me.Chart.SelectionStyle.Fill = New SolidBrush(Color.FromArgb(200, Color.CornflowerBlue))
            Me.Chart.ToolTip.Content = "X: {x} " & vbLf & "Y: {y}"
            Me.Chart.AxisX.LabelMax = True
            Me.Chart.AxisX.LabelMin = True
            AddHandler Me.Chart.MouseClick, AddressOf OnChartMouseClick
            AddHandler Me.Chart.Rendered, AddressOf OnChartRendered
            _pnlControlOptions = New FlowLayoutPanel() With {
                .FlowDirection = FlowDirection.LeftToRight,
                .Dock = DockStyle.Top,
                .Height = 30,
                .AutoScroll = True
            }
            _chbClear = New CheckBox() With {
                .Text = "Clear",
                .Checked = True,
                .Enabled = False,
                .AutoSize = True
            }
            AddHandler _chbClear.CheckedChanged, AddressOf OnClearChanged
            _cbSelectionMode = New ComboBox() With {
                .Width = 80,
                .Visible = ShowSelectionMode,
                .FlatStyle = FlatStyle.Flat,
                .DataSource = [Enum].GetValues(GetType(SelectionMode))
            }
            AddHandler _cbSelectionMode.SelectedIndexChanged, Sub(s, e)
                                                                  Me.SelectionMode = CType(_cbSelectionMode.SelectedValue, SelectionMode)
                                                                  OnSelectionModeChanged()
                                                              End Sub

            _pnlControlOptions.Controls.Add(_chbClear)
            _pnlControlOptions.Controls.Add(_cbSelectionMode)
            Dim isDateTime = items.First().[GetType]().GetProperty(Chart.BindingX).PropertyType = GetType(DateTime)
            _slider = New RangeSliderView(isDateTime) With {
                .Dock = DockStyle.Bottom
            }
            SetupRangeSlider()
            AddHandler Slider.RangeChanged, AddressOf RangeSliderView_RangeChanged
            Controls.Add(_pnlControlOptions)
            Controls.Add(Chart)
            Controls.Add(Slider)
        End Sub
#End Region
#Region "Private Methods"
#Region "Control Event Handlers"
        Protected Overrides Sub OnLoad(ByVal e As EventArgs)
            MyBase.OnLoad(e)
            _cbSelectionMode.SelectedIndex = CInt(SelectionMode)
            Slider.Visible = ShowRangeSlider
            Slider.Height = If(Slider.Height >= 80, 80, 40)
        End Sub

        Private Sub OnClearChanged()
            If _chbClear.Checked Then
                _selectedIndices.Clear()
                Slider.SetFullRange()
                OnSelectedPointsChanged()
            End If

            _chbClear.Enabled = Not _chbClear.Checked
        End Sub

        Private Sub OnChartRendered(ByVal sender As Object, ByVal e As RenderEventArgs)
            If _selecting Then
                e.Engine.SetStroke(Brushes.Red)
                Dim minX = Math.Min(_ptStart.X, _ptEnd.X)
                Dim minY = Math.Min(_ptStart.Y, _ptEnd.Y)
                Dim maxX = Math.Max(_ptStart.X, _ptEnd.X)
                Dim maxY = Math.Max(_ptStart.Y, _ptEnd.Y)
                e.Engine.DrawRect(minX, minY, maxX - minX, maxY - minY)
            End If
        End Sub

        Private Sub OnChartMouseClick(ByVal sender As Object, ByVal e As MouseEventArgs)
            Dim hitTest = Chart.HitTest(e.Location)

            If hitTest.Item IsNot Nothing AndAlso hitTest.Distance = 0 Then
                Dim currentSelected = hitTest.PointIndex

                If Not _selectedIndices.Contains(currentSelected) Then
                    _selectedIndices.Add(currentSelected)
                Else
                    _selectedIndices.Remove(currentSelected)
                End If

                OnSelectedPointsChanged()
            End If
        End Sub

        Private Sub OnChartMouseUp(ByVal sender As Object, ByVal e As MouseEventArgs)
            If _selecting Then
                _ptEnd = e.Location
                FindItemsInRectRange()
                _ptStart = Point.Empty
                _ptEnd = Point.Empty
                _selecting = False
                Chart.Capture = False
                Chart.Refresh()
            Else
                _ptStart = Point.Empty
                _ptEnd = Point.Empty
            End If
        End Sub

        Private Sub OnChartMouseMove(ByVal sender As Object, ByVal e As MouseEventArgs)
            If _ptStart <> Point.Empty Then
                _ptEnd = e.Location
                Chart.Refresh()
                _selecting = True
            End If
        End Sub

        Private Sub OnChartMouseDown(ByVal sender As Object, ByVal e As MouseEventArgs)
            _ptStart = e.Location
            Chart.Capture = True
        End Sub

        Private Sub OnSeriesSymbolRendering(ByVal sender As Object, ByVal e As RenderSymbolEventArgs)
            If _selectedIndices.Contains(e.Index) Then
                If Chart.SelectionStyle.Stroke IsNot Nothing Then e.Engine.SetStroke(Chart.SelectionStyle.Stroke)
                If Chart.SelectionStyle.StrokeWidth > 0 Then e.Engine.SetStrokeThickness(Chart.SelectionStyle.StrokeWidth)
                If Chart.SelectionStyle.Fill IsNot Nothing Then e.Engine.SetFill(Chart.SelectionStyle.Fill)
            End If
        End Sub

        Private Sub RangeSliderView_RangeChanged(ByVal sender As Object, ByVal e As EventArgs)
            If Not Slider.IsFullRange Then
                Chart.AxisX.Min = Slider.LowerValue
                Chart.AxisX.Max = Slider.UpperValue
            Else
                Chart.AxisX.Min = Double.NaN
                Chart.AxisX.Max = Double.NaN
            End If

            OnSelectedPointsChanged()
        End Sub
#End Region
        Private Sub OnSelectionModeChanged()
            RemoveHandler Chart.MouseClick, AddressOf OnChartMouseClick
            RemoveHandler Chart.MouseDown, AddressOf OnChartMouseDown
            RemoveHandler Chart.MouseMove, AddressOf OnChartMouseMove
            RemoveHandler Chart.MouseUp, AddressOf OnChartMouseUp

            Select Case SelectionMode
                Case SelectionMode.NormalSelect
                    AddHandler Chart.MouseClick, AddressOf OnChartMouseClick
                Case SelectionMode.RangeSelect
                    AddHandler Chart.MouseDown, AddressOf OnChartMouseDown
                    AddHandler Chart.MouseMove, AddressOf OnChartMouseMove
                    AddHandler Chart.MouseUp, AddressOf OnChartMouseUp
            End Select
        End Sub

        Private Sub OnSelectedPointsChanged()
            _chbClear.Checked = If(_selectedIndices.Count = 0, Slider.IsFullRange, False)
            Chart.Refresh()
            RaiseEvent SelectionChanged(Me, Nothing)
        End Sub

        Private Sub SetupRangeSlider()
            Dim values = GetChartXValues().ToList()
            Slider.Maximum = values.Max()
            Slider.Minimum = values.Min()
            Slider.SetFullRange()
        End Sub

        Private Sub FindItemsInRectRange()
            Dim startX = Chart.AxisX.ConvertBack(Math.Min(_ptStart.X, _ptEnd.X))
            Dim endX = Chart.AxisX.ConvertBack(Math.Max(_ptStart.X, _ptEnd.X))
            Dim startY = Chart.AxisY.ConvertBack(Math.Max(_ptStart.Y, _ptEnd.Y))
            Dim endY = Chart.AxisY.ConvertBack(Math.Min(_ptStart.Y, _ptEnd.Y))
            Dim xValues = GetChartXValues()
            Dim yValues = Chart.Series(0).GetValues(0)

            For idx As Integer = 0 To xValues.Count() - 1
                If (startX <= xValues(idx)) AndAlso (xValues(idx) <= endX) AndAlso (startY <= yValues(idx)) AndAlso (yValues(idx) <= endY) Then _selectedIndices.Add(idx)
            Next

            OnSelectedPointsChanged()
        End Sub

        Private Function GetChartXValues() As Double()
            Dim values = Chart.Series(0).GetValues(1)

            If values Is Nothing Then
                values = New Double(_items.Count() - 1) {}
                Dim idx As Integer = 0
                _items.ToList().ForEach(Sub(x)
                                            values(idx) = idx
                                            idx += 1
                                        End Sub)
            End If

            Return values
        End Function
#End Region
#Region "Public Methods"
        Public Function GetSelectedValues() As IEnumerable(Of Object)
            Dim items = Me._items.ToList()
            Return items.Where(Function(x) _selectedIndices.Contains(items.IndexOf(x)))
        End Function

        Public Function GetSliderRange() As Tuple(Of Object, Object)
            Dim values As Tuple(Of Object, Object) = Nothing
            If Slider.IsFullRange Then Return values

            If Slider.IsDateTimeRange AndAlso Not IsGroupedData Then
                values = New Tuple(Of Object, Object)(DateTime.FromOADate(Slider.LowerValue), DateTime.FromOADate(Slider.UpperValue))
            Else
                values = New Tuple(Of Object, Object)(Slider.LowerValue, Slider.UpperValue)
            End If

            Return values
        End Function

        Public Sub ChangeDataSource(ByVal items As IEnumerable(Of Object), ByVal Optional bindingX As String = "", ByVal Optional binding As String = "")
            Me._items = items
            Chart.DataSource = items
            Chart.BindingX = If(bindingX <> String.Empty, bindingX, Chart.BindingX)
            Chart.Series(0).Binding = If(binding <> String.Empty, binding, Chart.Series(0).Binding)
            Slider.Controls(0).Controls(1).Visible = Not IsGroupedData
            Slider.Height = If(Not IsGroupedData, 80, 40)
            SetupRangeSlider()
            _selectedIndices.Clear()
            OnSelectedPointsChanged()
        End Sub
#End Region
    End Class
End Namespace

File: Core/DataFilter/CS/DataFilterExplorer/Samples/Controls/FlexChartFilterView.cs
using C1.Chart;
using C1.Win.Chart;
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace DataFilterExplorer.Samples
{
    public partial class FlexChartFilterView : UserControl
    {
        #region fields
        private bool _isLoaded = false;
        private List<int> _selectedIndices;
        private bool _showSelectionMode = true, _showRangeSlider = false, _selecting;
        private Point _ptStart = Point.Empty, _ptEnd = Point.Empty;
        private IEnumerable<object> _items;
        private SelectionMode _selectionMode = SelectionMode.NormalSelect;
        private CheckBox _chbClear;
        private ComboBox _cbSelectionMode;
        protected FlowLayoutPanel _pnlControlOptions;

        TableLayoutPanel _tablePanel;

        #endregion

        #region Object Model

        public bool IsLoaded => _isLoaded && Slider.IsLoaded;

        public FlexChart Chart { get; }
        public RangeSliderView Slider { get; }

        [DefaultValue(SelectionMode.NormalSelect)]
        public SelectionMode SelectionMode
        {
            get
            {
                return _selectionMode;
            }
            set
            {
                if (_selectionMode != value)
                {
                    _selectionMode = value;
                    _cbSelectionMode.SelectedIndex = _cbSelectionMode.SelectedIndex == -1 ? -1 : (int)value;
                }
            }
        }

        [DefaultValue(true)]
        public bool ShowSelectionMode
        {
            get { return _showSelectionMode; }
            set
            {
                if (value != _showSelectionMode && _cbSelectionMode != null)
                {
                    _showSelectionMode = _cbSelectionMode.Visible = value;
                }
            }
        }

        [DefaultValue(false)]
        public bool ShowRangeSlider
        {
            get
            {
                return _showRangeSlider;
            }
            set
            {
                _showRangeSlider = value;
                if (this.IsHandleCreated)
                    Slider.Visible = value;
            }
        }
        public bool IsGroupedData
        {
            get { return _items.First().GetType() == typeof(GroupItem) || _items.First().GetType().IsSubclassOf(typeof(GroupItem)); }
        }

        public bool IsFilterApplied
        {
            get { return !_chbClear.Checked; }
        }

        public event EventHandler SelectionChanged;
        #endregion

        #region C'tor
        public FlexChartFilterView(IEnumerable<object> items,
            string bindingX,
            string binding,
            ChartType chartType = ChartType.Column)
        {
            InitializeComponent();

            Margin = new Padding(0);
            Padding = new Padding(0);

            _tablePanel = new TableLayoutPanel
            {
                Margin = new Padding(0),
                Padding = new Padding(0),
                RowCount = 3,
                Dock = DockStyle.Fill,
            };

            _selectedIndices = new List<int>();
            this._items = items;

            //Initialize Chart
            this.Chart = new FlexChart()
            {
                Margin = new Padding(0),
                Padding = new Padding(0),
                ChartType = chartType,
                DataSource = items,
                BindingX = bindingX,
                Binding = binding,
                Dock = DockStyle.Fill,
                //Margin = new Padding(0, 30, 0, 0)                
            };

            var ser = new Series();
            ser.SymbolRendering += OnSeriesSymbolRendering;
            this.Chart.Series.Add(ser);

            this.Chart.SelectionStyle.Stroke = Brushes.DarkBlue;
            this.Chart.SelectionStyle.StrokeWidth = 2;
            this.Chart.SelectionStyle.Fill = new SolidBrush(Color.FromArgb(200, Color.CornflowerBlue));
            this.Chart.ToolTip.Content = "X: {x} \nY: {y}";
            this.Chart.AxisX.LabelMax = this.Chart.AxisX.LabelMin = true;
            this.Chart.MouseClick += OnChartMouseClick;
            this.Chart.Rendered += OnChartRendered;

            _pnlControlOptions = new FlowLayoutPanel()
            {
                Margin = new Padding(0),
                Padding = new Padding(0),
                FlowDirection = FlowDirection.LeftToRight,
                //Dock = DockStyle.Top,
                Dock = DockStyle.Fill,
                //Height = 30,
                //AutoScroll = true,
            };

            _chbClear = new CheckBox()
            {
                Text = "Clear",
                Checked = true,
                Enabled = false,
                AutoSize = true,
            };
            _chbClear.CheckedChanged += (s, e) => OnClearChanged();

            _cbSelectionMode = new ComboBox()
            {
                Width = 80,
                Visible = ShowSelectionMode,
                FlatStyle = FlatStyle.Flat,
                DataSource = Enum.GetValues(typeof(SelectionMode)),
            };
            _cbSelectionMode.SelectedIndexChanged += (s, e) =>
            {
                this.SelectionMode = (SelectionMode)_cbSelectionMode.SelectedValue;
                OnSelectionModeChanged();
            };

            _pnlControlOptions.Controls.Add(_chbClear);
            _pnlControlOptions.Controls.Add(_cbSelectionMode);

            var isDateTime = items.First().GetType().GetProperty(Chart.BindingX).PropertyType == typeof(DateTime);

            Slider = new RangeSliderView(isDateTime)
            {
                Margin = new Padding(0),
                Padding = new Padding(0),
                //Dock = DockStyle.Bottom,                
                Dock = DockStyle.Fill
            };
            SetupRangeSlider();
            Slider.RangeChanged += RangeSliderView_RangeChanged;

            _pnlControlOptions.BorderStyle = BorderStyle.FixedSingle;
            Slider.BorderStyle = BorderStyle.FixedSingle;

            _tablePanel.ColumnCount = 1;
            _tablePanel.ColumnStyles.Add(new ColumnStyle() { SizeType = SizeType.Percent, Width = 100 });
            _tablePanel.RowStyles.Add(new RowStyle() { Height = 30, SizeType = SizeType.Absolute });
            _tablePanel.RowStyles.Add(new RowStyle() { SizeType = SizeType.Percent, Height = 100 });


            _tablePanel.Controls.Add(_pnlControlOptions, 0, 0);
            _tablePanel.Controls.Add(Chart, 0, 1);
            _tablePanel.Controls.Add(Slider, 0, 2);

            Controls.Add(_tablePanel);

            float scale = (float)DeviceDpi / (float)C1.Win.Util.DpiHelper.DEFAULT_DPI;
            _tablePanel.Scale(new System.Drawing.SizeF(scale, scale));

            //Controls.Add(_pnlControlOptions);
            //Controls.Add(Chart);
            //Controls.Add(Slider);
        }
        #endregion

        #region Private Methods

        #region Control Event Handlers
        protected override void OnLoad(EventArgs e)
        {
            base.OnLoad(e);
            _cbSelectionMode.SelectedIndex = (int)SelectionMode;
            Slider.Visible = ShowRangeSlider;
            Slider.Height = Slider.Height >= 80 ? 80 : 40;
            _tablePanel.RowStyles.Add(new RowStyle() { SizeType = SizeType.Absolute, Height = Slider.Visible ? Slider.Height : 0 });
            _isLoaded = true;
        }
        private void OnClearChanged()
        {
            if (_chbClear.Checked)
            {
                _selectedIndices.Clear();
                Slider.SetFullRange();
                OnSelectedPointsChanged();
            }
            _chbClear.Enabled = !_chbClear.Checked;
        }
        private void OnChartRendered(object sender, RenderEventArgs e)
        {
            if (_selecting)
            {
                e.Engine.SetStroke(Brushes.Red);
                var minX = Math.Min(_ptStart.X, _ptEnd.X);
                var minY = Math.Min(_ptStart.Y, _ptEnd.Y);
                var maxX = Math.Max(_ptStart.X, _ptEnd.X);
                var maxY = Math.Max(_ptStart.Y, _ptEnd.Y);
                e.Engine.DrawRect(minX, minY, maxX - minX, maxY - minY);
            }
        }

        private void OnChartMouseClick(object sender, MouseEventArgs e)
        {
            var hitTest = Chart.HitTest(e.Location);
            if (hitTest.Item != null && hitTest.Distance == 0)
            {
                var currentSelected = hitTest.PointIndex;
                if (!_selectedIndices.Contains(currentSelected))
                    _selectedIndices.Add(currentSelected);
                else
                    _selectedIndices.Remove(currentSelected);
                OnSelectedPointsChanged();
            }
        }

        private void OnChartMouseUp(object sender, MouseEventArgs e)
        {
            if (_selecting)
            {
                _ptEnd = e.Location;
                FindItemsInRectRange();
                _ptStart = Point.Empty;
                _ptEnd = Point.Empty;
                _selecting = false;

                Chart.Capture = false;
                Chart.Refresh();
            }
            else
            {
                _ptStart = Point.Empty;
                _ptEnd = Point.Empty;
            }
        }

        private void OnChartMouseMove(object sender, MouseEventArgs e)
        {
            if (_ptStart != Point.Empty)
            {
                _ptEnd = e.Location;
                Chart.Refresh();
                _selecting = true;
            }
        }

        private void OnChartMouseDown(object sender, MouseEventArgs e)
        {
            _ptStart = e.Location;
            Chart.Capture = true;
        }
        private void OnSeriesSymbolRendering(object sender, RenderSymbolEventArgs e)
        {
            if (_selectedIndices.Contains(e.Index))
            {
                if (Chart.SelectionStyle.Stroke != null)
                    e.Engine.SetStroke(Chart.SelectionStyle.Stroke);
                if (Chart.SelectionStyle.StrokeWidth > 0)
                    e.Engine.SetStrokeThickness(Chart.SelectionStyle.StrokeWidth);
                if (Chart.SelectionStyle.Fill != null)
                    e.Engine.SetFill(Chart.SelectionStyle.Fill);
            }
        }
        private void RangeSliderView_RangeChanged(object sender, EventArgs e)
        {
            if (!Slider.IsFullRange)
            {
                Chart.AxisX.Min = Slider.LowerValue;
                Chart.AxisX.Max = Slider.UpperValue;
            }
            else
            {
                Chart.AxisX.Min = Chart.AxisX.Max = double.NaN;
            }
            OnSelectedPointsChanged();
        }
        #endregion

        private void OnSelectionModeChanged()
        {
            Chart.MouseClick -= OnChartMouseClick;
            Chart.MouseDown -= OnChartMouseDown;
            Chart.MouseMove -= OnChartMouseMove;
            Chart.MouseUp -= OnChartMouseUp;

            switch (SelectionMode)
            {
                case SelectionMode.NormalSelect:
                    Chart.MouseClick += OnChartMouseClick;
                    break;
                case SelectionMode.RangeSelect:
                    Chart.MouseDown += OnChartMouseDown;
                    Chart.MouseMove += OnChartMouseMove;
                    Chart.MouseUp += OnChartMouseUp;
                    break;
            }
        }

        private void OnSelectedPointsChanged()
        {
            _chbClear.Checked = _selectedIndices.Count == 0 ? Slider.IsFullRange : false;
            Chart.Refresh();
            SelectionChanged?.Invoke(this, null);
        }

        private void SetupRangeSlider()
        {
            var values = GetChartXValues().ToList();
            Slider.Maximum = values.Max();
            Slider.Minimum = values.Min();
            Slider.SetFullRange();
        }
        private void FindItemsInRectRange()
        {
            var startX = Chart.AxisX.ConvertBack(Math.Min(_ptStart.X, _ptEnd.X));
            var endX = Chart.AxisX.ConvertBack(Math.Max(_ptStart.X, _ptEnd.X));
            var startY = Chart.AxisY.ConvertBack(Math.Max(_ptStart.Y, _ptEnd.Y));
            var endY = Chart.AxisY.ConvertBack(Math.Min(_ptStart.Y, _ptEnd.Y));

            var xValues = GetChartXValues();
            var yValues = Chart.Series[0].GetValues(0);

            for (int idx = 0; idx < xValues.Count(); idx++)
            {
                if ((startX <= xValues[idx]) && (xValues[idx] <= endX) && (startY <= yValues[idx]) && (yValues[idx] <= endY))
                    _selectedIndices.Add(idx);
            }
            OnSelectedPointsChanged();
        }
        private double[] GetChartXValues()
        {
            var values = Chart.Series[0].GetValues(1);
            if (values == null)
            {
                values = new double[_items.Count()];
                int idx = 0;
                _items.ToList().ForEach(x => values[idx] = idx++);
            }
            return values;
        }
        #endregion

        #region Public Methods
        public IEnumerable<object> GetSelectedValues()
        {
            var items = this._items.ToList();
            return items.Where(x => _selectedIndices.Contains(items.IndexOf(x)));
        }
        public Tuple<object, object> GetSliderRange()
        {
            Tuple<object, object> values = null;

            if (Slider.IsFullRange)
                return values;
            if (Slider.IsDateTimeRange && !IsGroupedData)
                values = new Tuple<object, object>(DateTime.FromOADate(Slider.LowerValue), DateTime.FromOADate(Slider.UpperValue));
            else
                values = new Tuple<object, object>(Slider.LowerValue, Slider.UpperValue);
            return values;
        }
        public void ChangeDataSource(IEnumerable<object> items, string bindingX = "", string binding = "")
        {
            Chart.DataSource = this._items = items;
            Chart.BindingX = bindingX != string.Empty ? bindingX : Chart.BindingX;
            Chart.Series[0].Binding = binding != string.Empty ? binding : Chart.Series[0].Binding;
            Slider.Controls[0].Controls[1].Visible = !IsGroupedData;
            Slider.Height = !IsGroupedData ? 80 : 40;
            SetupRangeSlider();
            _selectedIndices.Clear();
            OnSelectedPointsChanged();
        }
        #endregion
    }
}
